### PR TITLE
train method

### LINKED ISSUE #906: Tests are correct, but take too long
Test are unreasonably long, especially in the test_model.py. Adam suspects it has to do with the marginal_ll default. Let's investigate this.

### PR DESCRIPTION
**Pull request by galenxing**
Fixes #906 

Breaks out the train method into extensible object-oriented sequence.

### CODE REVIEW

File: scvi/model/_scanvi.py [hunk 1]
```diff
@@ -6,18 +6,21 @@
 import torch
 from anndata import AnnData
 from pandas.api.types import CategoricalDtype
-from sklearn.model_selection._split import _validate_shuffle_split
 
-from scvi import _CONSTANTS, settings
+from scvi import _CONSTANTS  # , settings
```
Comment (adamgayoso) [inline scvi/model/_scanvi.py:10]: flag

File: scvi/dataloaders/_data_splitting.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        self.adata = adata
+        self.train_size = float(train_size)
+        self.validation_size = validation_size
+        self.data_loader_kwargs = kwargs
+        self.use_gpu = use_gpu
+
+        self.train_idx, self.test_idx, self.val_idx = self.make_splits()
+
+    def make_splits(self):
+        if self.train_size > 1.0 or self.train_size <= 0.0:
```
Comment (adamgayoso) [inline scvi/dataloaders/_data_splitting.py:30]: let's also check that val_size is valid (in 0 to 1)

Comment (adamgayoso) [reply scvi/dataloaders/_data_splitting.py:30]: apparently it works to give val size a large integer right now

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/external/solo/_model.py [hunk 3]
```diff
@@ -185,17 +185,37 @@ def train(
...
+            training_plan=training_plan,
+            data_splitter=data_splitter,
+            max_epochs=max_epochs,
+            gpus=gpus,
             **kwargs,
         )
+        return runner()
 
-    def _set_training_plan(self, plan_class, plan_kwargs):
+    def _set_training_plan(self, plan_class=None, plan_kwargs=None):
```
Comment (adamgayoso) [inline scvi/external/solo/_model.py:218]: Remove

File: scvi/model/_scanvi.py [hunk 4]
```diff
@@ -386,139 +289,90 @@ def predict(
...
-        dl_kwargs.update(kwargs)
-
-        scanvi_train_dl = self._make_data_loader(
-            adata,
-            indices=indices_train,
-            shuffle=True,
-            data_loader_class=data_loader_class,
-            drop_last=3,
-            **dl_kwargs,
+        # if we have labeled cells, we want to subsample labels each epoch
```
Comment (adamgayoso) [inline scvi/model/_scanvi.py:348]: can be removed right?

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/model/_scanvi.py (hunk 4 above)
Comment (adamgayoso) [reply scvi/model/_scanvi.py:348]: referring to commented out code above

Comment (galenxing) [reply scvi/model/_scanvi.py:348]: yea im removing in the gpu pr

File: scvi/dataloaders/_data_splitting.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+        float, or None (default is 0.9)
+    validation_size
+        float, or None (default is None)
+    use_gpu
+        Which GPU to use
+    **kwargs
+        Keyword args for data loader. If adata has labeled data, data loader
+        class is :class:`~scvi.dataloaders.SemiSupervisedDataLoader`,
+        else data loader class is :class:`~scvi.dataloaders.AnnDataLoader`.
+    """
```
Comment (adamgayoso) [inline scvi/dataloaders/_data_splitting.py:71]: add examples

File: scvi/external/solo/_model.py [hunk 6]
```diff
@@ -12,18 +12,18 @@
...
+from scvi.model.base import BaseModelClass, TrainRunner, UnsupervisedTrainingMixin
 from scvi.modules import Classifier
 
 logger = logging.getLogger(__name__)
 
 LABELS_KEY = "_solo_doub_sim"
 
 
-class SOLO(BaseModelClass):
+class SOLO(UnsupervisedTrainingMixin, BaseModelClass):
```
Comment (adamgayoso) [inline scvi/external/solo/_model.py:26]: why does it need the mixin?

Comment (adamgayoso) [reply scvi/external/solo/_model.py:26]: it should only need it if you call super below, which seems like a fine option to me

File: scvi/model/base/_training_mixin.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+            If `True`, use the GPU if available. Will override the use_gpu option when initializing model
+        train_size
+            Size of training set in the range [0.0, 1.0].
+        validation_size
+            Size of the test set. If `None`, defaults to 1 - `train_size`. If
+            `train_size + validation_size < 1`, the remaining cells belong to a test set.
+        batch_size
+            Minibatch size to use during training.
+        plan_kwargs
+            Keyword args for model-specific Pytorch Lightning task. Keyword arguments passed to
```
Comment (adamgayoso) [inline scvi/model/base/_training_mixin.py:41]:
```suggestion
            Keyword args for :class:`~scvi.lightning.TrainingPlan`. Keyword arguments passed to
```

File: scvi/model/base/_trainrunner.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    data_splitter
+        initialized :class:`~scvi.dataloaders.SemiSupervisedDataSplitter` or
+        :class:`~scvi.dataloaders.DataSplitter`
+    max_epochs
+        max_epochs to train for
+    gpus
+        Which GPU to train on
+    trainer_kwargs
+        Extra kwargs for :class:`~scvi.lightning.Trainer`
+    """
```
Comment (adamgayoso) [inline scvi/model/base/_trainrunner.py:32]: examples

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/dataloaders/_data_splitting.py [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+        n_labeled_idx = len(self._labeled_indices)
+        n_unlabeled_idx = len(self._unlabeled_indices)
+
+        if n_labeled_idx != 0:
+            n_labeled_train, n_labeled_val = validate_data_split(
+                n_labeled_idx, self.train_size, self.validation_size
+            )
+            labeled_permutation = np.random.choice(
+                self._labeled_indices, len(self._labeled_indices), replace=False
+            )
```
Comment (adamgayoso) [inline scvi/dataloaders/_data_splitting.py:200]: use RandomState like in the main splitter

File: scvi/dataloaders/_data_splitting.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+            labeled_idx_train = []
+            labeled_idx_val = []
+
+        if n_unlabeled_idx != 0:
+            n_unlabeled_train, n_unlabeled_val = validate_data_split(
+                n_unlabeled_idx, self.train_size, self.validation_size
+            )
+            unlabeled_permutation = np.random.choice(
+                self._unlabeled_indices, len(self._unlabeled_indices)
+            )
```
Comment (adamgayoso) [inline scvi/dataloaders/_data_splitting.py:217]: randomstate

**Review by adamgayoso — APPROVED** [review]
LGTM

**Pull request merged**